File: intellij-plugin/src/main/kotlin/com/ezevillo/gitreview/domain/TimeoutClass.kt
package com.ezevillo.gitreview.domain

enum class InvocationClass {
    READ,
    LOCAL_MUTATION,
    NETWORK,
    SUPPORT_GIT,
}

const val READ_TIMEOUT_MS: Long = 15_000
const val LOCAL_MUTATION_TIMEOUT_MS: Long = 120_000
const val NETWORK_MUTATION_TIMEOUT_MS: Long = 300_000
const val SUPPORT_GIT_TIMEOUT_MS: Long = 30_000

private val LOCAL_MUTATION_VERBS = setOf(
    "finish", "save", "abort", "continue", "next", "prev",
    "clean", "forget", "compare", "walkthrough", "preview",
)

private val NETWORK_MUTATION_VERBS = setOf("start")

fun timeoutForClass(verb: String, args: List<String>): Long {
    if (verb in NETWORK_MUTATION_VERBS) return NETWORK_MUTATION_TIMEOUT_MS
    if (verb == "forget" && "--stale" in args) return NETWORK_MUTATION_TIMEOUT_MS
    if (verb in LOCAL_MUTATION_VERBS) return LOCAL_MUTATION_TIMEOUT_MS
    return READ_TIMEOUT_MS
}

fun timeoutMs(invocationClass: InvocationClass): Long = when (invocationClass) {
    InvocationClass.READ -> READ_TIMEOUT_MS
    InvocationClass.LOCAL_MUTATION -> LOCAL_MUTATION_TIMEOUT_MS
    InvocationClass.NETWORK -> NETWORK_MUTATION_TIMEOUT_MS
    InvocationClass.SUPPORT_GIT -> SUPPORT_GIT_TIMEOUT_MS
}

File: visualstudio-extension/src/GitReview.Domain/TimeoutClass.cs
namespace GitReview.Domain;

public enum InvocationClass
{
    Read,
    LocalMutation,
    Network,
    SupportGit,
}

public static class TimeoutClass
{
    public const long ReadTimeoutMs = 15_000;
    public const long LocalMutationTimeoutMs = 120_000;
    public const long NetworkMutationTimeoutMs = 300_000;
    public const long SupportGitTimeoutMs = 30_000;

    private static readonly HashSet<string> LocalMutationVerbs = new(StringComparer.Ordinal)
    {
        "finish", "save", "abort", "continue", "next", "prev",
        "clean", "forget", "compare", "walkthrough", "preview",
    };

    private static readonly HashSet<string> NetworkMutationVerbs = new(StringComparer.Ordinal)
    {
        "start",
    };

    public static long TimeoutForClass(string verb, IReadOnlyList<string> args)
    {
        if (NetworkMutationVerbs.Contains(verb)) return NetworkMutationTimeoutMs;
        if (verb == "forget" && args.Contains("--stale")) return NetworkMutationTimeoutMs;
        if (LocalMutationVerbs.Contains(verb)) return LocalMutationTimeoutMs;
        return ReadTimeoutMs;
    }

    public static long TimeoutMs(InvocationClass invocationClass) => invocationClass switch
    {
        InvocationClass.Read => ReadTimeoutMs,
        InvocationClass.LocalMutation => LocalMutationTimeoutMs,
        InvocationClass.Network => NetworkMutationTimeoutMs,
        InvocationClass.SupportGit => SupportGitTimeoutMs,
        _ => ReadTimeoutMs,
    };
}
